### ISSUE TITLE
Implement AsRawFd and FromRawFd for ReadDir

### ISSUE THREAD

**Issue by hvenev** [rust-lang/rust#56777]
Most unixes provide `dirfd()`, which can be used to get a file descriptor from `struct DIR *`.

The two errors are `EINVAL` (`dirfd` called on invalid pointer), which can't occur in correct code, and `ENOTSUP` (not supported by implementation), which doesn't occur on most implementations.

Most unixes also provide `fdopendir()`, which opens a file descriptor as `struct DIR *`. It fails if the file descriptor is invalid or if it fails to allocate memory.

**Comment by chloekek** [comment]
I need the FromRawFd impl so I’d like to work on this.

**Comment by chloekek** [comment]
With [I/O safety](https://github.com/rust-lang/rust/issues/87074) being stabilized this issue should also include implementing `AsFd`, `From<OwnedFd>`, and `From<ReadDir> for OwnedFd`.

**Comment by the8472** [comment]
I don't think we can provide `AsFd` due to the API requirements of `fdopendir()`. After #120426 we may be able to provide it on some platforms, but it won't be fully portable so we either have to hide it behind `cfg`s or make the API fallible.
